### ISSUE TITLE
I wish for better support for classes that delegate

### ISSUE THREAD

**Issue by HansMuller** [dart-lang/dartdoc#1161]
Documenting a class that delegates many parameters to another class is tedious. I've been copying the parameter dartdoc around which is error prone and brittle. What I wish for is a way to say the doc for some parameter in one class is the same as in another class. Something like this:

```
class Foo {   
  /// [same-doc-as Bar.foo]
  final int foo;
}
```

This situation came up repeatedly in https://github.com/flutter/flutter/pull/3950

Using one of the following approaches obviates copying but hurts readability:

```
/// Forwards to [Bar.foo]
final int foo;
```

or

```
/// See [Bar.foo]
final int foo;
```

**Comment by lrhn** [comment]
There is no common interface for `Foo` and `Bar` that declares `foo`?

I think the use-case of having multiple independent methods with the same behavior (same doc), but without a common interface declaring the method, is pretty rare.

**Comment by Hixie** [comment]
It happens all the time in Flutter because we have two parallel hierarchies. The RenderObject subclasses define the layout algorithms, and each one is wrapped by a RenderObjectWidget subclass. The two don't have any common superclasses. Their API surface is quite different (e.g. properties are mutable on the RenderObject side but immutable on the RenderObjectWidget side).

We also sometimes have features which form a common pattern, e.g. it is common for abstract classes to have a const constructor. We want the same (or substantially similar) text for each case of that pattern, but we don't want to have to duplicate the text everywhere. Right now, almost every abstract class const constructor has a copied-and-pasted version of the same paragraph. It would be much better if the docs could all just point to one place that has that definition so that if we update it once, it updates everywhere.

**Comment by Hixie** [comment]
The OP mentions https://github.com/flutter/flutter/pull/3950/files - search for "Whether to place first child at the start" in that PR for another example of where this comes up.

**Comment by lrhn** [comment]
A quick look at that code shows that the four places duplicating `scrollAnchor` also duplicates  `initalScrollOffset`, `scrollDirection`, `onScrollStart`, `onScroll` and `onScrollEnd` (maybe more, I stopped counting there).

That highly suggests that there is some interface in play, even if you haven't named it.

That suggests a solution: You can define the `scrollAnchor` (with documentation* in an interface and implement that interface in all four classes. That should inherit the documentation. Maybe also the other shared members. You can make it a private interface if you just want the documentation sharing (and the classes happen to be in the same library).

**Comment by Hixie** [comment]
The problem is we wouldn't want that interface to be visible publicly, yet we wouldn't be able to mark it private (since it'd need to be used in multiple files). We could, I suppose, have a file that we import into the library but never export, though that would be very unusual in our code base. It would make the docs weird, though, as now these classes would all claim to implement this secret non-private non-exported interface, they would all claim these members were inherited even though they aren't really, and it would also mean we were changing the semantics to fit the documentation, which seems backwards.

Also, it wouldn't help with the common case described in https://github.com/dart-lang/dartdoc/issues/1161#issuecomment-242131715.

**Comment by lrhn** [comment]
That is true. I can see that would probably be nice to be able to just write `//@inherit(Foo.bar)` (or whatever) as the documentation, even if `Foo.bar` is somewhere else. You have to keep the same parameter names etc., and it's all or nothing, not "import".

**Comment by Hixie** [comment]
That (having to match the signatures) wouldn't help for the cases described in https://github.com/dart-lang/dartdoc/issues/1161#issuecomment-242131715 (namely constructors, and RenderObject mutable fields -> RenderObjectWidget immutable fields).

Come to think of it, we also have other cases where lots of fields have common blobs of text. The whole comment can't be copied, but it would be really nice to be able to import individual paragraphs so we only have to write them once.

**Comment by Hixie** [comment]
I've filed https://github.com/dart-lang/dartdoc/issues/1264 on the latter issue.

**Comment by astashov** [comment]
So, if I'd like to fix this issue, how would you want to see it? Maybe add ability to mark paragraphs by labels (like

```
|label: foobar|
Here are paragraphs,
lots of them

And more

And more
|label-end|
```

and then use them like:

```
It will be inserted

|insert: foobar|

right above
```

or something like that

**Comment by Hixie** [comment]
This particular bug is more about delegation than about macros, which your last comment looks like a solution for. That would be #1264. I'll comment there.

**Comment by astashov** [comment]
Yeah, but the problem of delegation could also be solved by macros. You'd just write something like:

```
class Foo extends Bar {   
  /// |insert: barfoo|
  final int foo;
}
```

I mean the question is - do we really need both? Generic macros/templating system and also more narrow use case - delegating docs.

**Comment by Hixie** [comment]
I agree that if we had macros that might take care of this one as well.

**Comment by astashov** [comment]
Should we then close this in favor of https://github.com/dart-lang/dartdoc/issues/1264? @devoncarew ?

**Comment by devoncarew** [comment]
I'm not sure we want a macro (or macro like) system. I think `@example` is good for a specific use-case; I would lean towards introducing another `@` dartdoc tag to handle this use case - that of borrowing docs from another symbol.

The currently supported doc tags are:
- `/// @nodoc`
- `/// {@example PATH [region=NAME] [lang=NAME]}`

I think we'd want syntax like:

```dart
/// @docsfrom [foo.bar]
```

Though I don't know the best name for the annotation. `@docsfrom`? `@borrowdocs`? `@inherit`?

**Comment by Hixie** [comment]
@devoncarew `@docsfrom` would handle this use case but not the ones in #1264.

**Comment by devoncarew** [comment]
Re the use case in that other issue, would the example annotation help there? That can textually include the content from files.

**Comment by Hixie** [comment]
If there's an `#include` analogue (which does markdown processing on the included text), and we don't mind having a directory somewhere with dozens of these files, then sure. The docs for `@example` aren't super clear to me (what's the `lang` bit for? What is the value of the `region` argument mutating `PATH`?), so I'm not sure how good a fit it would be.

Most of the includes we would do are a few lines long, e.g. the boilerplate we have for every abstract `const` constructor.

I don't really mind what the solution is. I have scheduled some time in the new year to examine using `@example`.

**Comment by devoncarew** [comment]
It doesn't sound like `@example` (as-is) would work for your use-case. @astashov, it does sound like this issue could be solved by #1264. It may still be nice to have a more semantic way to forward docs, but not as critical.

**Comment by astashov** [comment]
Code for #1264 was merged and #1264 was closed, can we close this one as well?
